src/core/comm/allgather_thread_local.cc

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "coll.h"
+#include "legion.h"
+
+namespace legate {
+namespace comm {
+namespace coll {
+
+using namespace Legion;
+extern Logger log_coll;
+
+int collAllgatherLocal(const void* sendbuf,
+                       int sendcount,
+                       CollDataType sendtype,
+                       void* recvbuf,
+                       int recvcount,
+                       CollDataType recvtype,
+                       CollComm global_comm)
+{
+  assert(recvcount == sendcount);
+  assert(sendtype == recvtype);
+
+  int total_size = global_comm->global_comm_size;
+
+  int sendtype_extent = collGetDtypeSize(sendtype);
+  int recvtype_extent = collGetDtypeSize(recvtype);
+  assert(sendtype_extent == recvtype_extent);
+
+  int global_rank = global_comm->global_rank;
+
+  void* sendbuf_tmp = NULL;
+
+  // MPI_IN_PLACE
+  if (sendbuf == recvbuf) {
+    sendbuf_tmp = (void*)malloc(sendtype_extent * sendcount);
+    assert(sendbuf_tmp != NULL);
+    memcpy(sendbuf_tmp, recvbuf, sendtype_extent * sendcount);
+  } else {
+    sendbuf_tmp = const_cast<void*>(sendbuf);
```

Comment:
it'd be more tasteful if you didn't have to use `const_cast`.

Reply:
There is no way to avoid it. I tried to set the `sendbuf_tmp `to `const void*`, but later, when I do pointer arithmetic and `free(sendbuf_tmp)`, I will need to do the `const_cast`.